agent/auto-config/auto_config.go

```diff
@@ -101,138 +56,61 @@ func WithOverrides(overrides ...config.Source) Option {
 // then we will need to add some locking here. I am deferring that for now
 // to help ease the review of this already large PR.
 type AutoConfig struct {
-	config          *config.RuntimeConfig
-	builderOpts     config.BuilderOpts
-	logger          hclog.Logger
-	directRPC       DirectRPC
-	tlsConfigurator *tlsutil.Configurator
-	autoConfigData  string
-	waiter          *lib.RetryWaiter
-	overrides       []config.Source
-}
-
-func flattenOptions(opts []Option) options {
-	var flat options
-	for _, opt := range opts {
-		opt(&flat)
-	}
-	return flat
+	// These fields do NOT need protection by the mutex
+	// They are immutable after new AutoConfig creation
+	// and write operations are not done on them or
+	// those other objects are them self concurrency safe
+	builderOpts config.BuilderOpts
+	logger      hclog.Logger
+	directRPC   DirectRPC
+	waiter      *lib.RetryWaiter
+	overrides   []config.Source
+	certMonitor CertMonitor
+
+	// l mainly is protecting the cancel and autoConfigData
+	l sync.Mutex
```

Comment:
I think currently we are not using `l` to lock anything here? I also think that we are not using the `AutoConfig` struct concurrently anywhere right now?

Reply:
Yep, I was going to but then I offloaded most of the concurrency to the CertMonitor

Reply:
Removed the mutex